l2-contracts/contracts/ConsensusRegistry.sol

```diff
@@ -0,0 +1,278 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.20;
+
+import {Ownable2Step} from "@openzeppelin/contracts/access/Ownable2Step.sol";
+import {EfficientCall} from "@matterlabs/zksync-contracts/l2/system-contracts/libraries/EfficientCall.sol";
+
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @title ConsensusRegistry
+/// @dev Manages consensus nodes and committees for the L2 consensus protocol,
+/// owned by a governance protocol. Nodes act as both validators and attesters,
+/// each playing a distinct role in the consensus process. This contract facilitates
+/// the rotation of validator and attester committees, which represent a subset of nodes
+/// expected to actively participate in the consensus process during a specific time window.
+contract ConsensusRegistry is Ownable2Step {
+    // An array to keep track of node owners.
+    address[] public nodeOwners;
+    // A map of node owners => nodes.
+    mapping(address => Node) public nodes;
```

Comment:
Currently `remove` and `add` function have `O(nodeOwners.length)` complexity with big constant overhead. 

- `changeX` functions are cheap though
- `setValidatorCommittee`/`setAttesterCommittee` very expensive.

Proposal to fix it:
1) Redesign storage layout. Instead of storing `nodeOwners` and `nodes` storage values (array and mapping), you can store one linked list. The example can be found on the [Safe multisig](https://github.com/safe-global/safe-smart-account/blob/main/contracts/base/OwnerManager.sol#L19). 
- We will store `mapping(address => NodeWithPointerToNext) public nodes;` where `NodeWithPointerToNext` is

```
struct NodeWithPointerToNext {
      Node node;
      address pointerToNext;
}
```

now you can store nodes by key of `node owner address`. The node will point to the next node in the linked list.

Access by `node owner address` - O(1), without need to traverse `nodeOwners` when want to remove the value from the array so:
- `remove` complexity will drop to O(1).

Reply:
Nice idea. I didn’t think such optimization are needed, since it’s an L2 contract. 

Let’s break it down:

* `add` isn’t O(N) but rather O(1) if we’re not considering the validations, which are unrelated to this optimization (they will still require to iterate)
* `setValidatorCommittee`/`setAttesterCommittee` complexity won’t change 
* We’ll need to store the root node, for allowing iterations 

So in the overall,

* pros: reduce `remove` runtime complexity from O(N) to O(1)
* cons: higher code complexity

Anything I’m missing?

Reply:
> Nice idea. I didn’t think such optimization are needed, since it’s an L2 contract.

L2 contract execution isn't free. The storage is expensive on L2 too.

> `add` isn’t O(N) but rather O(1) if we’re not considering the validations, which are unrelated to this optimization (they will still require to iterate)

On Slack, I provided the idea of how to optimize the validation. But in fact, a part of validation is iterating over `nodeOwners` that my proposal fixes.

> `setValidatorCommittee`/`setAttesterCommittee` complexity won’t change

It is a separate issue, that we will fix

> We’ll need to store the root node, for allowing iterations

By design of the data structure

> So in the overall, pros: reduce remove runtime complexity from O(N) to O(1), cons: higher code complexity

Together with saving hashes of the validator pubkeys we will reduce `add` complexity from O(N) to O(1). And eliminate possible issues with permanent contract freeze due to inefficient storage writes (when you can't execute the transaction because it consumes too much execution or data)

Reply:
If we're to tradeoff computation for storage by storing the public key hashes, there will still be a need to iterate over these for comparison. So I'm not sure how `add` complexity will be reduced from O(N) to O(1).

Reply:
Right, I assumed that we don't want to introduce more storage, but we can just maintain the hash on the map instead of the node struct.